flink-tests/src/test/java/org/apache/flink/runtime/metrics/JobManagerMetricsITCase.java

```diff
@@ -0,0 +1,221 @@
+	private static final String TASK_SLOTS_AVAILABLE = "taskSlotsAvailable";
+	private static final String TASK_SLOTS_TOTAL = "taskSlotsTotal";
+	private static final String NUM_REGISTERED_TASK_MANAGERS = "numRegisteredTaskManagers";
+	private static final String NUM_RUNNING_JOBS = "numRunningJobs";
+
+	private static CountDownLatch invokeLatch;
+
+	private MiniClusterClient clusterClient;
+	private JobGraph jobGraph;
+
+	@ClassRule
+	public static final MiniClusterResource MINI_CLUSTER_RESOURCE = new MiniClusterResource(
+		new MiniClusterResourceConfiguration.Builder()
+			.setConfiguration(getConfiguration())
+			.setNumberTaskManagers(1)
+			.setNumberSlotsPerTaskManager(1)
+			.build());
+
+	@Before
+	public void setUp() throws Exception {
+		invokeLatch = new CountDownLatch(1);
+
+		Assume.assumeTrue(
+			"ClusterClient is not an instance of MiniClusterClient",
+			MINI_CLUSTER_RESOURCE.getClusterClient() instanceof MiniClusterClient);
+
+		clusterClient = (MiniClusterClient) MINI_CLUSTER_RESOURCE.getClusterClient();
+		clusterClient.setDetached(true);
+
+		jobGraph = new JobGraph();
+
+		final JobVertex vertex = new JobVertex("testVertex");
+		vertex.setInvokableClass(NoOpBlockingInvokable.class);
+		jobGraph.addVertex(vertex);
+
+		clusterClient.submitJob(jobGraph, ClassLoader.getSystemClassLoader());
+		invokeLatch.await(60, TimeUnit.SECONDS);
+		waitForJob();
+	}
+
+	@Test
+	public void testJobManagerMetrics() throws Exception {
+		assertEquals(1, TestReporter.OPENED_REPORTERS.size());
+		TestReporter reporter = TestReporter.OPENED_REPORTERS.iterator().next();
+
+		List<String> expectedPatterns = getExpectedPatterns();
+
+		Collection<String> gaugeNames = reporter.getGauges().values();
+
+		for (String expectedPattern : expectedPatterns) {
+			boolean found = false;
+			for (String gaugeName : gaugeNames) {
+				if (gaugeName.matches(expectedPattern)) {
+					found = true;
+				}
+			}
+			if (!found) {
+				fail(String.format("Failed to find gauge [%s] in registered gauges [%s]",
+					expectedPattern, gaugeNames));
+			}
+		}
+
+		for (Map.Entry<Gauge<?>, String> entry : reporter.getGauges().entrySet()) {
+			if (entry.getValue().contains(TASK_SLOTS_AVAILABLE)) {
+				assertEquals(0L, entry.getKey().getValue());
+			} else if (entry.getValue().contains(TASK_SLOTS_TOTAL)) {
+				assertEquals(1L, entry.getKey().getValue());
+			} else if (entry.getValue().contains(NUM_REGISTERED_TASK_MANAGERS)) {
+				assertEquals(1L, entry.getKey().getValue());
+			} else if (entry.getValue().contains(NUM_RUNNING_JOBS)) {
+				assertEquals(1L, entry.getKey().getValue());
+			}
+		}
+	}
+
+	private static Configuration getConfiguration() {
+		Configuration configuration = new Configuration();
+		configuration.setString(REPORTERS_LIST, "test_reporter");
+		configuration.setString("metrics.reporter.test_reporter.class", TestReporter.class.getName());
+		return configuration;
+	}
+
+	private void waitForJob() throws Exception {
```

Comment:
This method should be called `waitForJobRunning()`; however this entire logic can be implemented more nicely with a [BlockerSync](https://github.com/apache/flink/blob/master/flink-test-utils-parent/flink-test-utils-junit/src/main/java/org/apache/flink/core/testutils/BlockerSync.java) anyway. The UDF would wait until the test thread calls `block`, and then you reverse this for unblocking the thread.

This method could then be removed.